File: Bootloader/Grub.hs
{- dfsbuild: CD image builder
Copyright (c) 2006 John Goerzen
Please see COPYRIGHT for more details
-}

module Bootloader.Grub where
import Utils
import MissingH.Cmd
import MissingH.Path
import System.Posix.Files
import System.Posix.Directory
import MissingH.Path.Glob
import MissingH.ConfigParser
import MissingH.Path.FilePath
import Data.List
import Actions.ConfigFiles

grub_eltorito env =
    do im "Installing bootloader: Grub raw eltorito (no HD emulation)"
       grub_generic env
       return (["-b", "boot/grub/stage2_eltorito", "-no-emul-boot",
                "-boot-load-size", "1", "-boot-info-table"],
               (\_ -> return ()))

grub_hd env =
    do im "Installing bootloader: Grub with eltorito HD emulation"
       grub_generic env
       safeSystem "cp" ["-r", (targetdir env) ++ "/boot", workbootdir]
       safeSystem "rm" ["-f", workbootdir ++ "/grub/stage2_eltorito"]
       bracketCWD (wdir env) $
          safeSystem "tar" ["-zcpf", "boot.tar.gz", "boot"]
       safeSystem "mkbimage" ["-f", workboottar, "-t", "hd", "-s", "ext2",
                              "-d", wdir env]
       rename "hd.image" $ (targetdir env) ++ "/boot/hd.image"
       return (["-b", "boot/hd.image", "-hard-disk-boot", "-c",
                "boot/boot.catalog"],
               (\_ -> return ()))
    where workbootdir = (wdir env) ++ "/boot"
          workboottar = (wdir env) ++ "/boot.tar.gz"

grub_generic env =
    do createDirectory (targetdir env ++ "/boot/grub") 0o755
	   -- since etch (Debian 4.0) grub files are located in /usr/lib instead of /lib
       grubfiles_pre_etch <- glob "/lib/grub/*/*"
       grubfiles_since_etch <- glob "/usr/lib/grub/*/*"
       safeSystem "cp" $ ["-rv"] ++ grubfiles_pre_etch ++ grubfiles_since_etch ++
         [targetdir env ++ "/boot/grub/"]
       menuText <- grubMenu env
       writeFile (targetdir env ++ "/boot/grub/menu.lst") menuText

       -- Help text not presently references
       writeFile (targetdir env ++ "/boot/grub/help.lst") helpText
       

grubMenu env  =
    do newkerns <- glob $ targetdir env ++ "/boot/vmlinu*"
       kerntext <- mapM kern (sort $ newkerns)
       return $ 
          case get (cp env) (defaultArch env) "grubconfig" of
            Left _ -> ""
            Right line -> line ++ "\n"
          ++ "color cyan/blue blue/light-gray\n"
          ++ (concat kerntext)
          ++ fake "."
          ++ fake (getidstring env)
    where fake s = "title " ++ s ++ "\ncolor cyan/blue blue/light-gray\n"
          kern x = do initrd <- getinitrdname env x
                      rootdev <- getrootdevname env x
                      return $ 
                        "title  Boot " ++ (snd . splitFileName $ x) ++ "\n"
                        ++ "kernel /boot/" ++ (snd . splitFileName $ x) ++ 
                               " root=" ++ rootdev ++ "\n"
                        ++ "initrd /boot/" ++ initrd ++ "\n"
                        ++ "boot\n"

helpText = "pager on\n\
\title Basic Booting Info\n\
\cat /opt/dfsruntime/dfs.html/booting.html.txt\n\
\ \n\
\title Selecting CD-ROM device\n\
\cat /opt/dfsruntime/dfs.html/dfsboot-selcd.html.txt\n\
\ \n\
\title About This CD\n\
\cat /opt/dfsruntime/buildinfo\n\
\ \n\
\title .\n\
\color cyan/black blue/light-gray\n\
\ \n\
\title Return to main menu...\n\
\configfile /boot/grub/menu.lst\n"

File: bootloaders/grub.ml
(* arch-tag: Grub support
* Copyright (c) 2004 John Goerzen
*)

open Unix;;
open Shellutil;;
open Dfsutils;;
open Archsupport;;

let grub_generic cp target entryline =
  mkdir (target ^ "/boot/grub") 0o755;
  run "cp" ("-rv" :: glob ["/usr/lib/grub/*/*"] @ [target ^ "/boot/grub/"]);
  let sd = open_out (target ^ "/boot/grub/menu.lst") in
  if cp#has_option (getarch()) "grubconfig" then begin
    output_string sd (get cp "grubconfig");
    output_string sd "\n";
  end;
  output_string sd "color cyan/blue blue/light-gray\n";
  let newkerns = glob [target ^ "/boot/vmlinu*"] in
  let os s = output_string sd (s ^ "\n") in
  let fake s = os ("title " ^ s ^ "\ncolor cyan/blue blue/light-gray") in
  List.iter (fun x ->
    os ("title  Boot " ^ (Filename.basename x));
    os ("kernel /boot/" ^ (Filename.basename x) ^ " root=/dev/ram0 " ^ \
      (Bootloaderutil.getrdparam target));
    os (entryline);
    os ("boot\n");
  ) newkerns;

  (*
  fake ".";
  os ("title Help/Information Menu\nconfigfile /boot/grub/help.lst\n");
  *)
  fake ".";
  fake (Configfiles.getidstring cp);

  Pervasives.close_out sd;

  let sd2 = open_out (target ^ "/boot/grub/help.lst") in
  output_string sd2 "color cyan/black blue/light-gray
pager on
title Basic Booting Info
cat /opt/dfsruntime/dfs.html/booting.html.txt

title Selecting CD-ROM device
cat /opt/dfsruntime/dfs.html/dfsbood-selcd.html.txt

title About This CD
cat /opt/dfsruntime/buildinfo

title .
color cyan/black blue/light-gray

title Return to main menu...
configfile /boot/grub/menu.lst
";
  Pervasives.close_out sd2;;
  
let grub_eltorito cp target =
  grub_generic cp target "initrd /opt/dfsruntime/initrd.dfs";
  installrd_cramfs target;
  (["-b"; "boot/grub/stage2_eltorito"; "-no-emul-boot";
   "-boot-load-size"; "1"; "-boot-info-table"],
   fun cp wdir target isoname -> ());;

let grub_hd cp workdir target =
  grub_generic cp target "initrd /boot/initrd.dfs";
  installrd_cramfs target;
  let workbootdir = workdir ^ "/boot" in
  let workboottar = workdir ^ "/boot.tar.gz" in
  run "cp" ["-r"; target ^ "/boot"; workbootdir];
  rm ~force:true (workbootdir ^ "/grub/stage2_eltorito");
  run "cp" [target ^ "/opt/dfsruntime/initrd.dfs"; workbootdir];
  run "sh" ["-c"; "cd " ^ workdir ^ "; tar -zcpf boot.tar.gz boot"];
  run "mkbimage" ["-f"; workboottar; "-t"; "hd"; "-s"; "ext2"; "-d"; workdir];
  Unix.rename "hd.image" (target ^ "/boot/hd.image");
  (["-b"; "boot/hd.image"; "-hard-disk-boot"; "-c"; "boot/boot.catalog"],
   fun cp wdir target isoname -> ());;
